### PR TITLE
Initial commit of autodiff cookbook

### PR DESCRIPTION
**Pull request by alexbw**
Work by mattjj@ and myself. Hopefully one of many cookbooks to come!
You can view and run the notebook while reviewing using this link: http://colab.research.google.com/github/google/jax/blob/autodiff-cookbook/notebooks/autodiff_cookbook.ipynb

TODO:
- [x] add an explanation of complex number support
- [x] show that `jacfwd` and `jacrev` work with pytrees
- [x] write in complete sentences as needed
- [x] maybe add linearize-once Gauss-Newton (or decide not to)
- [x] hint at a Part 2 with more advanced stuff coming soon

### CODE REVIEW

File: notebooks/autodiff_cookbook.ipynb [hunk 1]
```diff
@@ -0,0 +1,1062 @@
...
+        "\n",
+        "Mathematically, given a function $f : \\mathbb{R}^n \\to \\mathbb{R}^m$, the Jacobian matrix of $f$ evaluated at an input point $x \\in \\mathbb{R}^n$, denoted $\\partial f(x)$, is often thought of as a matrix in $\\mathbb{R}^m \\times \\mathbb{R}^n$:\n",
+        "\n",
+        "$\\qquad \\partial f(x) \\in \\mathbb{R}^{m \\times n}$.\n",
+        "\n",
+        "But we can also think of $\\partial f(x)$ as a linear map, which maps the tangent space of the domain of $f$ at the point $x$ (which is just another copy of $\\mathbb{R}^n$) to the tangent space of the codomain of $f$ at the point $f(x)$ (a copy of $\\mathbb{R}^m$):\n",
+        "\n",
+        "$\\qquad \\partial f(x) : \\mathbb{R}^n \\to \\mathbb{R}^m$.\n",
+        "\n",
+        "This map is called the [pushforward map](https://en.wikipedia.org/wiki/Pushforward_(differential) of $f$ at $x$. The Jacobian matrix is just the matrix for this linear map in a standard basis.\n",
```
Comment (jekbradbury) [inline notebooks/autodiff_cookbook.ipynb]:
```suggestion
        "This map is called the [pushforward map](https://en.wikipedia.org/wiki/Pushforward_(differential)) of $f$ at $x$. The Jacobian matrix is just the matrix for this linear map in a standard basis.\n",
```

File: notebooks/autodiff_cookbook.ipynb [hunk 2]
```diff
@@ -0,0 +1,1062 @@
...
+        "# Differentiate `loss` with respect to the first positional argument:\n",
+        "W_grad = grad(loss, argnums=0)(W, b)\n",
+        "print('W_grad', W_grad)\n",
+        "\n",
+        "# Since argnums=0 is the default, this does the same thing:\n",
+        "W_grad = grad(loss)(W, b)\n",
+        "print('W_grad', W_grad)\n",
+        "\n",
+        "# But we can choose different values too:\n",
+        "b_grad = grad(loss, 1)(W, b)\n",
```
Comment (hawkinsp) [inline notebooks/autodiff_cookbook.ipynb]: You might wish to always pass `argnums` by name here (and below). Consistency is a good thing. If you wish to switch to positional, perhaps comment that you are doing so.

File: notebooks/autodiff_cookbook.ipynb [hunk 3]
```diff
@@ -0,0 +1,1062 @@
...
+      ]
+    },
+    {
+      "metadata": {
+        "id": "3RpbiasHGD3X",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "Memory cost doesn't scale with the depth of the computation. FLOP cost about 2x evaluating just the function, with the cost of $(x, v) \\mapsto (x, \\partial f(x) v)$ has about the same cost to evaluate as $f$.\n",
```
Comment (hawkinsp) [inline notebooks/autodiff_cookbook.ipynb]: Missing verb in the "FLOP cost" sentence.

File: notebooks/autodiff_cookbook.ipynb [hunk 4]
```diff
@@ -0,0 +1,1062 @@
...
+        "\n",
+        "But think of building a Jacobian matrix: one column at a time. Efficient for \"tall\" Jacobians but inefficient for \"wide\" Jacobians.\n",
```
Comment (hawkinsp) [inline notebooks/autodiff_cookbook.ipynb]: Remove the colon after "matrix".

Comment (hawkinsp) [reply notebooks/autodiff_cookbook.ipynb]: "Efficient for ..." is a sentence fragment. Rephrase.

File: notebooks/autodiff_cookbook.ipynb [hunk 5]
```diff
@@ -0,0 +1,1062 @@
...
+      },
+      "cell_type": "markdown",
+      "source": [
+        "### Vector-Jacobian products (VJPs, aka reverse-mode autodiff)\n",
+        "\n",
+        "Where forward-mode gives us back a function for evaluating Jacobian-vector products, which we can then use to build Jacobian matrices one column at a time, reverse-mode is a way to get back a function for evaluating vector-Jacobian products (equivalently Jacobian-transpose-vector products), which we can use to build Jacobian matrices one row at a time.\n",
+        "\n",
+        "#### VJPs in math\n",
+        "\n",
+        "Let's again consider a function $f : \\mathbb{R}^n \\to \\mathbb{R}^m$\n",
```
Comment (hawkinsp) [inline notebooks/autodiff_cookbook.ipynb]: Missing period at end of sentence.

File: notebooks/autodiff_cookbook.ipynb [hunk 6]
```diff
@@ -0,0 +1,1062 @@
...
+        "Starting from our notation for JVPs, the notation for VJPs is pretty simple:\n",
+        "\n",
+        "$\\qquad (x, v) \\mapsto v \\partial f(x)$,\n",
+        "\n",
+        "where here $v$ is an element of the cotangent space of $f$ at $x$ (isomorphic to another copy of $\\mathbb{R}^m$). When being rigorous, we should think of $v$ as a linear map $v : \\mathbb{R}^m \\to \\mathbb{R}$, and when we write $v \\partial f(x)$ we mean function composition $(v \\circ \\partial f)(x)$. But in the common case we can identify it with a vector in $\\mathbb{R}^m$ and use the two almost interchageably, just like we might sometimes flip between \"column vectors\" and \"row vectors\" without much comment.\n",
```
Comment (hawkinsp) [inline notebooks/autodiff_cookbook.ipynb]: Remove "here".

File: notebooks/autodiff_cookbook.ipynb [hunk 7]
```diff
@@ -0,0 +1,1062 @@
...
+    {
+      "metadata": {
+        "id": "PSL1TciM6vvI",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "### Jacobian-Matrix and Matrix-Jacobian products\n",
+        "\n",
+        "Oh, that's interesting. No more loops?"
```
Comment (hawkinsp) [inline notebooks/autodiff_cookbook.ipynb]: This needs more text to fit with the style of the previous text. Why is it interesting? Complete sentences please!

Comment (hawkinsp) [reply notebooks/autodiff_cookbook.ipynb]: If nothing else, use the opportunity to pitch `vmap`.

File: notebooks/autodiff_cookbook.ipynb [hunk 8]
```diff
@@ -0,0 +1,1062 @@
...
+    {
+      "metadata": {
+        "id": "MXFEFBDz6vvL",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "### The implementation of `jacfwd` and `jacrev`\n",
+        "\n",
+        "Cool, cool, we got this. *Autograd couldn't do this.*"
```
Comment (hawkinsp) [inline notebooks/autodiff_cookbook.ipynb]: Same here. Why is it cool?

File: notebooks/autodiff_cookbook.ipynb [hunk 9]
```diff
@@ -0,0 +1,1062 @@
...
+      },
+      "cell_type": "code",
+      "source": [
+        "from jax import hessian as builtin_hessian\n",
+        "\n",
+        "def our_hessian(f):\n",
+        "  # Forward-over-reverse calculation of the Hessian.\n",
+        "  # We could also do jacfwd(jacfwd(W)), jacrev(jacrev(W)) or jacrev(jacfwd(W)).\n",
+        "  # They're all equally valid. Try them out and time them!\n",
+        "  return jacfwd(jacrev(f))\n",
```
Comment (hawkinsp) [inline notebooks/autodiff_cookbook.ipynb]: Our usual formulation also sticks a `jit()` around this, you may wish to do the same or add text observing that you can so do.

**Review by hawkinsp — APPROVED** [review]
This looks really good, except the last few sections need a bit more text.

File: notebooks/autodiff_cookbook.ipynb (hunk 3 above)
Comment (mattjj) [reply notebooks/autodiff_cookbook.ipynb]: Yeah I think these are incomplete sentences that never got finished! Thanks for catching.

**Comment by mattjj** [comment]
It's okay, googlebot; @jekbradbury works for Google.

File: notebooks/autodiff_cookbook.ipynb (hunk 2 above)
Comment (mattjj) [reply notebooks/autodiff_cookbook.ipynb]: Added a comment.

File: notebooks/autodiff_cookbook.ipynb (hunk 3 above)
Comment (mattjj) [reply notebooks/autodiff_cookbook.ipynb]: Added words!

File: notebooks/autodiff_cookbook.ipynb (hunk 5 above)
Comment (mattjj) [reply notebooks/autodiff_cookbook.ipynb]: Done, thanks.

File: notebooks/autodiff_cookbook.ipynb (hunk 4 above)
Comment (mattjj) [reply notebooks/autodiff_cookbook.ipynb]: Done.

File: notebooks/autodiff_cookbook.ipynb (hunk 7 above)
Comment (mattjj) [reply notebooks/autodiff_cookbook.ipynb]: Added words!

File: notebooks/autodiff_cookbook.ipynb (hunk 8 above)
Comment (mattjj) [reply notebooks/autodiff_cookbook.ipynb]: Added comments to the end.

File: notebooks/autodiff_cookbook.ipynb (hunk 9 above)
Comment (mattjj) [reply notebooks/autodiff_cookbook.ipynb]: Added a version of this.

**Comment by mattjj** [comment]
I'm in a bitter fight with googlebot.

File: notebooks/autodiff_cookbook.ipynb [hunk 10]
```diff
@@ -0,0 +1,1252 @@
...
+    "    return jacfun\n",
+    "\n",
+    "assert np.allclose(builtin_jacfwd(f)(W), our_jacfwd(f)(W)), 'Incorrect forward-mode Jacobian results!'"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Interestingly, [Autograd](https://github.com/hips/autograd) couldn't do this. Our [implementation of reverse-mode `jacobian` in Autograd](https://github.com/HIPS/autograd/blob/96a03f44da43cd7044c61ac945c483955deba957/autograd/differential_operators.py#L60) had to pull back one vector at a time with an outer-loop `map`. Pushing one vector at a time through the computattion is much less efficient than batching it all together with `vmap`."
```
Comment (hawkinsp) [inline notebooks/autodiff_cookbook.ipynb]: Typo: computattion

File: notebooks/autodiff_cookbook.ipynb [hunk 11]
```diff
@@ -0,0 +1,1252 @@
...
+    "    try:\n",
+    "        if x < 3:\n",
+    "            return 2 * x ** 3\n",
+    "        else:\n",
+    "            raise ValueError\n",
+    "    except ValueError:\n",
+    "        return np.pi * x\n",
+    "\n",
+    "y, f_vjp = vjp(f, 4.)\n",
+    "print(f_vjp(1.))"
```
Comment (hawkinsp) [inline notebooks/autodiff_cookbook.ipynb]: Should there be an application of `jit` here?

File: notebooks/autodiff_cookbook.ipynb [hunk 12]
```diff
@@ -0,0 +1,1252 @@
...
+   "metadata": {},
+   "source": [
+    "## Complex numbers and differentiation"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "JAX is great at complex numbers and differentiation. To support both holomorphic and non-holomorphic differentiation, JAX follows [Autograd's convention](https://github.com/HIPS/autograd/blob/master/docs/tutorial.md#complex-numbers) for encoding complex derivatives.\n",
```
Comment (hawkinsp) [inline notebooks/autodiff_cookbook.ipynb]: You might link to "holomorphic function" on Wikipedia the first time you mention it, if, as I do, you need a refresher in complex analysis.

**Review by hawkinsp — APPROVED** [review]
LGTM!

**Comment by alexbw** [comment]
GOOGLE BOT COME ON

**Comment by jekbradbury** [comment]
I'm pretty sure it's all my fault for editing the notebook in the GitHub web UI, where my commit email was `@gmail.com` rather than `@google.com`... Sorry about that!

**Pull request merged**